src/mapping/query/function/split.rs

```diff
@@ -0,0 +1,247 @@
+fn regex_from_map(mut map: BTreeMap<String, Value>) -> Result<(regex::Regex, bool)> {
+    let pattern = map
+        .remove("pattern")
+        .ok_or_else(|| "Field is not a regular expression".to_string())?;
+
+    let flags = match map.remove("flags") {
+        None => Value::from(Vec::<Value>::new()),
+        Some(flags) => flags,
+    };
+
+    match (flags, pattern) {
+        (Value::Array(ref flags), Value::Bytes(ref bytes)) => {
+            let (global, insensitive, multi_line) =
+                flags
+                    .iter()
+                    .fold((false, false, false), |(g, i, m), flag| match flag {
+                        v if v == G_FLAG.deref() => (true, i, m),
+                        v if v == I_FLAG.deref() => (g, true, m),
+                        v if v == M_FLAG.deref() => (g, i, true),
+                        _ => (g, i, m),
+                    });
+
+            let pattern = String::from_utf8_lossy(&bytes);
+            let regex = regex::RegexBuilder::new(&pattern)
+                .case_insensitive(insensitive)
+                .multi_line(multi_line)
+                .build()
+                .map_err(|err| format!("invalid regex {}", err))?;
+            Ok((regex, global))
+        }
+        _ => Err("Field regular expression is not a valid string".to_string()),
+    }
+}
+
+impl Function for SplitFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let string = {
+            let bytes = required!(ctx, self.path, Value::Bytes(v) => v);
+            String::from_utf8_lossy(&bytes).into_owned()
+        };
+
+        let limit = optional!(ctx, self.limit, Value::Integer(i) => i)
+            .map(|limit| {
+                if limit < 0 {
+                    Err("limit is not a positive number".to_string())
+                } else {
+                    Ok(limit as usize)
+                }
+            })
+            .transpose()?;
+
+        let to_value = |iter: Box<dyn Iterator<Item = &str>>| {
+            Value::Array(
+                iter.map(|sub| Value::Bytes(sub.to_string().into()))
+                    .collect(),
+            )
+        };
+
+        match self.pattern.execute(ctx)? {
+            Value::Bytes(path) => {
+                let pattern = String::from_utf8_lossy(&path).into_owned();
+                Ok(match limit {
+                    Some(limit) => to_value(Box::new(string.splitn(limit, &pattern))),
+                    None => to_value(Box::new(string.split(&pattern))),
+                })
+            }
+            Value::Map(pattern) => {
+                let (regex, _global) = regex_from_map(pattern)?;
+                Ok(match &limit {
+                    Some(ref limit) => to_value(Box::new(regex.splitn(&string, *limit))),
+                    None => to_value(Box::new(regex.split(&string))),
+                })
+            }
+            _ => Err("invalid pattern".to_string()),
+        }
```

Comment:
I think this is the part we'd eventually want to remove into a separate `impl Function for LazyRegex`, or whatever newtype we'll end up with.